src/controllers/cube-handler.ts

```diff
@@ -0,0 +1,514 @@
+            case FactTableColumnType.Measure:
+                compositeKey.push(`"${field.columnName}"`);
+                factIdentifiers.push(field);
+                measureColumn = field;
+                break;
+            case FactTableColumnType.NoteCodes:
+                notesCodeColumn = field;
+                break;
+            case FactTableColumnType.DataValues:
+                dataValuesColumn = field;
+                break;
+        }
+        return `"${field.columnName}" ${field.columnDatatype}`;
+    });
+
+    if (!notesCodeColumn) {
+        throw Error(`No column representing notes codes was found`);
+    }
+    if (!dataValuesColumn) {
+        throw Error(`No column representing data was found`);
+    }
+    logger.info('Creating fact table in cube');
+    try {
+        await quack.exec(
+            `CREATE TABLE ${FACT_TABLE_NAME} (${factTableDef.join(', ')}, PRIMARY KEY (${compositeKey.join(', ')}));`
+        );
+    } catch (err) {
+        logger.error(`Failed to create fact table in cube: ${err}`);
+        await quack.close();
+        throw new Error(`Failed to create fact table in cube: ${err}`);
+    }
+
+    // Find all the fact tables for the given revision
+    let allFactTables: FactTable[] = [];
+    if (endRevision.revisionIndex > 0) {
+        // If we have a revision index we start here
+        const validRevisions = dataset.revisions.filter(
+            (rev) => rev.revisionIndex <= endRevision.revisionIndex && rev.revisionIndex > 0
+        );
+        allFactTables = validRevisions.flatMap((revision) => revision.factTables);
+    } else {
+        // If we don't have a revision index we need to find the previous revision to this one that does
+        allFactTables = allFactTables.concat(endRevision.factTables);
+        const validRevisions = dataset.revisions.filter(
+            (rev) => rev.createdAt < endRevision.createdAt && rev.revisionIndex > 0
+        );
+        allFactTables = validRevisions.flatMap((revision) => revision.factTables);
+    }
+
+    // Process all the fact tables
+    logger.debug('Loading all fact tables in to database');
+    try {
+        for (const factTable of allFactTables.sort((ftA, ftB) => ftA.uploadedAt.getTime() - ftB.uploadedAt.getTime())) {
+            logger.info(`Loading fact table data for fact table ${factTable.id}`);
+            const factTableFile = await getFileImportAndSaveToDisk(dataset, factTable);
+            const updateQuery =
+                `UPDATE ${FACT_TABLE_NAME} SET "${dataValuesColumn.columnName}"=update_table."${dataValuesColumn.columnName}", ` +
+                `"${notesCodeColumn.columnName}"=(CASE ${FACT_TABLE_NAME}."${notesCodeColumn.columnName}" = NULL THEN 'r' ELSE concat(${FACT_TABLE_NAME}."${notesCodeColumn.columnName}"', ',r') END) ` +
+                `FROM update_table WHERE ${setupFactTableUpdateJoins(factIdentifiers)} ` +
+                `AND update_table."${notesCodeColumn.columnName}" LIKE '%r';`;
```

Comment:
So I did try that...  I felt this was easier to read and also allowed me to quickly replace lines when I was coding.  I think we need to agree a style for handling long SQL in code and in a way which is easy enough to read.  I don't mind concatenating string for larger SQL queries if its easy to read.

Reply:
This particular query isn't too bad because it's already split across multiple lines - I'm thinking something like:

```
const sql = `
  SELECT x.*, y.*
  FROM x
  JOIN y ON x.foo = y.foo
  WHERE y.bar = 'baz';
`;
```

with breaking out the select items to multiple lines when it's a big list is fine.

Basically it's easier to scroll vertically than horizontally, and the whitespace in the query from indentation shouldn't matter to the db.